File: tools/tf-types-ts/src/core/actor-id.ts
import type { ActorType } from "../generated/_common.js";
import { sha256 } from "@noble/hashes/sha2";
import { toHex } from "./crypto.js";

export class ActorIdParseError extends Error {}

export const ACTOR_TYPES: readonly ActorType[] = [
  "human",
  "agent",
  "device",
  "service",
  "site",
  "organization",
  "relay",
  "plugin",
  "process",
  "tool",
  "model-provider",
  "policy-engine",
  "proof-anchor",
  "emergency-authority",
];

const ACTOR_TYPE_SET: ReadonlySet<string> = new Set<string>(ACTOR_TYPES);

export interface ParsedActorId {
  readonly type: ActorType;
  readonly path: string;
  readonly raw: string;
}

/**
 * Parse an actor URI of the form `tf:actor:<type>:<path>`.
 *
 * Per RFC 3986 the scheme is case-insensitive but we require the canonical
 * lower-case form; the colon-delimited type must be one of the 14 types
 * enumerated in TF-0002; the path segment is required and non-empty.
 */
export function parseActorId(s: string): ParsedActorId {
  if (typeof s !== "string") throw new ActorIdParseError(`expected string, got ${typeof s}`);
  const parts = splitScheme(s);
  if (!parts) throw new ActorIdParseError(`expected tf:actor:<type>:<path>, got ${JSON.stringify(s)}`);
  if (parts.kind !== "actor") {
    throw new ActorIdParseError(`expected scheme 'tf:actor:', got 'tf:${parts.kind}:'`);
  }
  if (!ACTOR_TYPE_SET.has(parts.typeSegment)) {
    throw new ActorIdParseError(`unknown actor type: ${parts.typeSegment}`);
  }
  if (parts.path.length === 0) throw new ActorIdParseError("actor id path is empty");
  return { type: parts.typeSegment as ActorType, path: parts.path, raw: s };
}

export function formatActorId(p: { type: ActorType; path: string }): string {
  if (!ACTOR_TYPE_SET.has(p.type)) throw new ActorIdParseError(`unknown actor type: ${p.type}`);
  if (!p.path) throw new ActorIdParseError("actor id path is empty");
  return `tf:actor:${p.type}:${p.path}`;
}

export function actorIdEquals(a: string, b: string): boolean {
  try {
    const pa = parseActorId(a);
    const pb = parseActorId(b);
    return pa.type === pb.type && pa.path === pb.path;
  } catch {
    return false;
  }
}

/**
 * Derive the canonical actor URI for a peer identified by an ed25519 public
 * key. The URI is bound to the key by construction: `tf:actor:process:key/<hex>`
 * where `<hex>` is the lowercase hex prefix of `sha256(ident_pub)` truncated
 * to 8 bytes (16 hex chars). This is the cryptographic identity the daemon
 * passes to AgentGuard; callers can additionally accept a `peer_hint` as a
 * self-claimed alias, but it is advisory only — never used for authority.
 */
export function derivePeerActor(identPub: Uint8Array): string {
  if (identPub.length !== 32) {
    throw new ActorIdParseError(`derivePeerActor expects 32-byte ed25519 public key, got ${identPub.length}`);
  }
  const digest = sha256(identPub);
  const thumbprint = toHex(digest.slice(0, 8));
  return `tf:actor:process:key/${thumbprint}`;
}

export function splitScheme(s: string): { kind: string; typeSegment: string; path: string } | null {
  if (!s.startsWith("tf:")) return null;
  const rest = s.slice(3);
  const firstColon = rest.indexOf(":");
  if (firstColon < 0) return null;
  const kind = rest.slice(0, firstColon);
  const remainder = rest.slice(firstColon + 1);
  const secondColon = remainder.indexOf(":");
  if (secondColon < 0) return null;
  const typeSegment = remainder.slice(0, secondColon);
  const path = remainder.slice(secondColon + 1);
  return { kind, typeSegment, path };
}

File: crates/tf-types/src/actor_id.rs
//! Actor-URI parser and formatter mirroring `tools/tf-types-ts/src/core/actor-id.ts`.

use crate::generated::common::ActorType;

#[derive(Debug, thiserror::Error, PartialEq, Eq)]
pub enum ActorIdParseError {
    #[error("expected tf:actor:<type>:<path>, got {0:?}")]
    MalformedScheme(String),
    #[error("expected scheme 'tf:actor:', got 'tf:{0}:'")]
    WrongKind(String),
    #[error("unknown actor type: {0}")]
    UnknownType(String),
    #[error("actor id path is empty")]
    EmptyPath,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct ParsedActorId {
    pub actor_type: ActorType,
    pub path: String,
    pub raw: String,
}

pub const ACTOR_TYPE_STRS: &[&str] = &[
    "human", "agent", "device", "service", "site", "organization",
    "relay", "plugin", "process", "tool", "model-provider",
    "policy-engine", "proof-anchor", "emergency-authority",
];

pub fn parse_actor_id(s: &str) -> Result<ParsedActorId, ActorIdParseError> {
    let parts = split_scheme(s).ok_or_else(|| ActorIdParseError::MalformedScheme(s.to_owned()))?;
    if parts.kind != "actor" {
        return Err(ActorIdParseError::WrongKind(parts.kind.to_owned()));
    }
    let actor_type = parse_actor_type(parts.type_segment)
        .ok_or_else(|| ActorIdParseError::UnknownType(parts.type_segment.to_owned()))?;
    if parts.path.is_empty() {
        return Err(ActorIdParseError::EmptyPath);
    }
    Ok(ParsedActorId {
        actor_type,
        path: parts.path.to_owned(),
        raw: s.to_owned(),
    })
}

pub fn format_actor_id(actor_type: &ActorType, path: &str) -> Result<String, ActorIdParseError> {
    if path.is_empty() {
        return Err(ActorIdParseError::EmptyPath);
    }
    Ok(format!("tf:actor:{}:{}", actor_type_to_str(actor_type), path))
}

pub fn actor_id_equals(a: &str, b: &str) -> bool {
    match (parse_actor_id(a), parse_actor_id(b)) {
        (Ok(pa), Ok(pb)) => pa.actor_type == pb.actor_type && pa.path == pb.path,
        _ => false,
    }
}

pub(crate) struct SchemeParts<'a> {
    pub kind: &'a str,
    pub type_segment: &'a str,
    pub path: &'a str,
}

pub(crate) fn split_scheme(s: &str) -> Option<SchemeParts<'_>> {
    let rest = s.strip_prefix("tf:")?;
    let first = rest.find(':')?;
    let kind = &rest[..first];
    let remainder = &rest[first + 1..];
    let second = remainder.find(':')?;
    let type_segment = &remainder[..second];
    let path = &remainder[second + 1..];
    Some(SchemeParts { kind, type_segment, path })
}

pub(crate) fn parse_actor_type(s: &str) -> Option<ActorType> {
    Some(match s {
        "human" => ActorType::Human,
        "agent" => ActorType::Agent,
        "device" => ActorType::Device,
        "service" => ActorType::Service,
        "site" => ActorType::Site,
        "organization" => ActorType::Organization,
        "relay" => ActorType::Relay,
        "plugin" => ActorType::Plugin,
        "process" => ActorType::Process,
        "tool" => ActorType::Tool,
        "model-provider" => ActorType::ModelProvider,
        "policy-engine" => ActorType::PolicyEngine,
        "proof-anchor" => ActorType::ProofAnchor,
        "emergency-authority" => ActorType::EmergencyAuthority,
        _ => return None,
    })
}

pub(crate) fn actor_type_to_str(t: &ActorType) -> &'static str {
    match t {
        ActorType::Human => "human",
        ActorType::Agent => "agent",
        ActorType::Device => "device",
        ActorType::Service => "service",
        ActorType::Site => "site",
        ActorType::Organization => "organization",
        ActorType::Relay => "relay",
        ActorType::Plugin => "plugin",
        ActorType::Process => "process",
        ActorType::Tool => "tool",
        ActorType::ModelProvider => "model-provider",
        ActorType::PolicyEngine => "policy-engine",
        ActorType::ProofAnchor => "proof-anchor",
        ActorType::EmergencyAuthority => "emergency-authority",
    }
}
